src/posix/platform/multicast_routing.cpp

```diff
@@ -0,0 +1,563 @@
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+
+#include "posix/platform/multicast_routing.hpp"
+
+#include <assert.h>
+#include <chrono>
+#include <net/if.h>
+#include <netinet/icmp6.h>
+#include <netinet/in.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+#if __linux__
+#include <linux/mroute6.h>
+#else
+#error "Multicast Routing feature is not ported to non-Linux platforms yet."
+#endif
+
+#include <openthread/backbone_router_ftd.h>
+
+#include "core/common/logging.hpp"
+
+namespace ot {
+namespace Posix {
+
+void MulticastRoutingManager::Init(otInstance *aInstance)
+{
+    otBackboneRouterSetMulticastListenerCallback(aInstance,
+                                                 &MulticastRoutingManager::HandleBackboneMulticastListenerEvent, this);
+}
+
+void MulticastRoutingManager::HandleBackboneMulticastListenerEvent(void *                                 aContext,
+                                                                   otBackboneRouterMulticastListenerEvent aEvent,
+                                                                   const otIp6Address *                   aAddress)
+{
+    static_cast<MulticastRoutingManager *>(aContext)->HandleBackboneMulticastListenerEvent(
+        aEvent, static_cast<const Ip6::Address &>(*aAddress));
+}
+
+void MulticastRoutingManager::HandleBackboneMulticastListenerEvent(otBackboneRouterMulticastListenerEvent aEvent,
+                                                                   const Ip6::Address &                   aAddress)
+{
+    switch (aEvent)
+    {
+    case OT_BACKBONE_ROUTER_MULTICAST_LISTENER_ADDED:
+        Add(aAddress);
+        break;
+    case OT_BACKBONE_ROUTER_MULTICAST_LISTENER_REMOVED:
+        Remove(aAddress);
+        break;
+    }
+}
+
+void MulticastRoutingManager::Enable(void)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(!IsEnabled());
+
+    SuccessOrExit(error = InitMulticastRouterSock());
+
+exit:
+    otLogResultPlat(error, "MulticastRoutingManager: %s", __FUNCTION__);
+}
+
+void MulticastRoutingManager::Disable(void)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(IsEnabled());
+
+    FinalizeMulticastRouterSock();
+
+exit:
+    otLogResultPlat(error, "MulticastRoutingManager: %s", __FUNCTION__);
+}
+
+void MulticastRoutingManager::Add(const Ip6::Address &aAddress)
+{
+    assert(mListenerSet.find(aAddress) == mListenerSet.end());
+    mListenerSet.insert(aAddress);
```

Comment:
why update `mListenerSet` regardless of `IsEnabled()`?

Reply:
I am assuming that it's possible to for SBBR to add multicast listener (at least in the future). So the current implementation does not assume that Multicast Listener is always added when it's Primary. But the `MulticastRoutingManager` is only enabled when it's Primary.

```
Spec 9.4.8.3.1:

Upon a successful registration of a Multicast Listener, the BBR sends out a BMLR.ntf
multicast on the Backbone Link interface to propagate all successfully processed multicast
registration entries to Secondary BBRs on the same Thread Network.
```

Note that the logic is same as before when using smcroute.

Reply:
> It may require additional efforts for this optional feature far beyond update the listener set here comparatively.
Yes. But I expect SBBR will also add Multicast Listener just like PBBR does. The difference is that MulticastRoutingManager is not enabled for SBBR. So, whether or not `MulticastRoutingManager` is enabled, it tries to maintain `mListenerSet` so that it's exactly same as the Multicast Listeners Table.

> Besides, one concern is there might be some inconsistence between mListenerSet here and PBBR's multicast listener table if there is some primary/secondary state switch, might bring assert in Line 109.

Yes, this was my concern too. Actually the purpose of adding the address to `mListenerSet` even if `MulticastRoutingManager` is not enabled is to **make sure mListenerSet and Multicast Listeners Table are exactly same**.